Add exception handling to Dataset context manager exit

darcymason opened the pull request:
#### Describe the changes
Related to issue #2168.  Adds a 'path' to the object where an error occurred, if using Dataset in a context manager `with` statement.

This is a first-pass rough proof of principle.
* requires some introspection of stack frames to try to determine the source, with some faith in finding which object the error point was operating on (here just use `self` from local vars).
* first commit just trapping for an error when setting an existing or new data element
* would need to work out different types of errors and how to determine which object information is useful to present
* first commit needs Python >= 3.11 to use `add_notes` for exceptions.  For Py 3.10, maybe do something like `tag_in_exception` does to add the message.
* credit:  `path_to` patterned closely to a json example from a Raymond Hettinger talk.  Uses `match`/`case` which is available Python 3.10+
* I've added this to Dataset, but could make a stand-alone context manager.  `Dataset` is nice because if `FileDataset`, the filename is reported.

Also
* happened to see and fix a DS_numpy docstring


Looking for feedback before pursuing further.  Of course, so far this only helps a user locate where it happened (super-charged version of `tag_in_exception`).  Still need to address more meaningful info about conversion errors, etc.


#### Tasks
- [x] (some) Unit tests added that reproduce the issue or prove feature is working
- [x] (in progress) Fix or feature added
- [ ] Code typed and mypy shows no errors
- [ ] Documentation updated (if relevant)
  - [ ] No warnings during build
  - [ ] Preview link (CircleCI -> Artifacts -> `doc/_build/html/index.html`)
- [ ] Unit tests passing and overall coverage the same or better

darcymason commented on src/pydicom/dataset.py:
```diff
@@ -440,6 +443,35 @@ def __exit__(
+            match (filename, frame.f_code.co_name):
+                case ("filewriter.py", "_correct_ambiguous_vr_element"):
+                    elem = frame.f_locals.get("elem")
+                case ("dataelem.py" | "sequence.py" | "dataset.py", func):
```
I had a new commit in progress where I have done so ... actually rearranged things a bit and is simpler, but also uses `_` more.  I had been leaving `func` because I thought more discriminating cases would be needed.

scaramallion commented on src/pydicom/dataset.py:
```diff
@@ -440,6 +443,49 @@ def __exit__(
+        if not hasattr(exc_val, "__notes__") or all(
+            n != note for n in exc_val.__notes__
+        ):
+            exc_val.add_note(note)
```
If `elem` is `None` then an empty note gets added here.

darcymason replied:
Fixed

scaramallion commented:
I really like this approach, but have you considered creating a generic context manager instead of adding this to the context manager within `Dataset` itself? That way it could be added inside the relevant parts of `dcmread()` and `dcmwrite()` (and anywhere else required) while still producing the same outcome.
```python
class add_exception_notes:
     def __init__(...): ...
     def __enter__(...): ...
     def __exit__(...): ...

def dcmread(...):
    ...
    with add_exception_notes:
        dataset  = read_partial(...)
```
Perhaps it could out-source its behaviour in `__exit__()` to a separate function as well, allowing you to customise the response as needed.

darcymason commented on src/pydicom/dataset.py:
```diff
@@ -440,6 +443,35 @@ def __exit__(
+            else:
+                new_elem = filename == "dataelem.py"
+
+        exc_val.add_note(note)
```
Fixed

darcymason commented on src/pydicom/dataset.py:
```diff
@@ -3673,3 +3719,67 @@ def _tsyntax_encoding(self) -> tuple[bool, bool] | tuple[None, None]:
 )
+
+
+def path_to(target, node) -> str | None:
```
Done

darcymason commented on src/pydicom/dataset.py:
```diff
@@ -3673,3 +3719,67 @@ def _tsyntax_encoding(self) -> tuple[bool, bool] | tuple[None, None]:
+                    prefix = (
+                        f"FileDataset(filename='{node.filename}')"
+                        if hasattr(node, "filename")
+                        else ""
```
I've reworked the code, I think it is a bit cleaner but have a look.

darcymason commented:
> have you considered creating a generic context manager instead of adding this to the context manager within `Dataset` itself? That way it could be added inside the relevant parts of `dcmread()` and `dcmwrite()` (and anywhere else required) while still producing the same outcome.

I considered a generic one, but I hoped to encourage getting the full path from the root dataset (so filename is often included) - very useful when processing a large number of files.   BTW, in latest commits, I did add a `with dataset` strategically to `dcmwrite` to catch most errors automatically, so much less need for the user to add it.  Could look into `dcmread` too.

scaramallion commented:
> I hoped to encourage getting the full path from the root dataset (so filename is often included) 

That could be done by passing kwargs with the filename to the context manager, but whichever method works

darcymason commented:
Okay, some general comments, getting close to complete from my point of view at least.

* I think having the full path to the object is really helpful for nested structures.
* New catches can be added quickly, just adding another `case` statement to capture a different code point.  I just added a check for a sequence index error in a few lines.
* it is Python >=3.11 only at the moment, because of the new "notes" capability.  Python 3.10 is EOL in a little over a year, so maybe not too bad ... it is only a helper function, so can point people with complex errors to re-running a Python version higher.
* The notes are seen when running from a command line, but at the moment, I don't see them in VS Code debug ... I'm assuming that will be added in some VS Code update.
*  We could instead do `raise ... from` as `tag_in_exception` did.  One downside is then we get repeated errors if the user does a `with ds` on top of the ones I've added internally.
* I replaced internal `tag_in_exception` calls with this, but left the function itself for backwards compatibility if users had used it; We could see how this goes and perhaps deprecate it at some later time.

mrbean-bremen commented on src/pydicom/dataset.py:
```diff
@@ -3673,3 +3728,82 @@ def _tsyntax_encoding(self) -> tuple[bool, bool] | tuple[None, None]:
+            for tag, dataelem in items:
+                if (path := _path_to(target, dataelem)) is not None:
+                    break
+            else:  # for-else = "no break", i.e. target not found
```
it's is a testimony to how not intuitive this is, if you have to explain a standard language feature 😁

scaramallion commented:
Can you add tests for the missing coverage?
